Generator seed

mh-nguyen712 opened the pull request:
### Checks to be done before submitting your PR
- [ ] `python3 -m pytest tests/` runs successfully.
- [x] `black .` runs successfully.
- [x] `make html` runs successfully (in the `docs/` directory).
- [x] Updated docstrings related to the changes (as applicable).
- [x] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

Hi deepinv team. I add a new feature to `PhysicsGenerator` class. Each class will have now a `rng` attribute for random number generator for reproducibility, as suggested in #243 . Moreover, we can also specify the `seed` in each generation step. I also fixed the `ProductConvolutionBlurGenerator` for generating a batch of eigen-PSF, which is not supported previously, and enhance the SVD quality. 
An another small feature is that I created an equivalent of `numpy.random.choice` in torch, which allows `PhysicsGenerator` from inpainting and mri to take torch.Generator instead of numpy.Generator, make it more uniform and avoid some overhead of switching between torch and numpy, mayby @Andrewwango  can take a look?

Don't hesitate to make any comments!

Andrewwango commented:
Hi @minh-hai-nguyen-712, thank you for the PR! I will review it when I return from annual leave w/c Mon 26th.

Andrewwango commented on deepinv/optim/fixed_point.py:
```diff
@@ -242,7 +242,6 @@ def forward(self, *args, compute_metrics=False, x_gt=None, **kwargs):
             range(self.max_iter),
             disable=(not self.verbose or not self.show_progress_bar),
         ):
-
```
Why are there lots of changed files with newlines removed?

minh-hai-nguyen-712 replied:
It's was due to black formatter. Also, as I made some improvements on SpaceVaryingBlurGenerator, which lead to some neccesary changes in SpaceVaryingBlur physics. The changes is mainly  about the documentation.

minh-hai-nguyen-712 replied:
And I forgot to turn off my autopep8 formatter so sometimes it changes `lambda x: ....` into `def func(x): ...`

Andrewwango replied:
Please can we remove the extra newlines changes as they aren't relevant to this PR?

minh-hai-nguyen-712 replied:
The extra changes came from when I merged this branch with main. I'll remove it as you suggested

Andrewwango commented on deepinv/physics/generator/base.py:
```diff
@@ -26,41 +29,75 @@ class PhysicsGenerator(nn.Module):
+                rng.device == device
+            ), "The random generator is not on the same device as the Physic Generator"
+            self.rng = rng
+        self.initial_random_state = self.rng.get_state()
```
Can we have a `reset_rng` method that sets the rng back to the initial state? That way, we can have reproducibility in epochs for #202 (provided that `shuffle=False` in `DataLoader`). And add a test to test this.

minh-hai-nguyen-712 replied:
Good idea. I'll try to do this as soon as possible.

Andrewwango replied:
```python
from torch.utils.data import DataLoader
import torch
from deepinv.tests.dummy_datasets.datasets import DummyCircles
from deepinv.physics import Inpainting
from deepinv.physics.generator import BernoulliSplittingMaskGenerator
from deepinv.utils import plot

# Sample dataset
img_size = (3, 8, 8)
dataset = DummyCircles(5, imsize=img_size)
loader = DataLoader(dataset, shuffle=False)

# Random inpainting physics
physics = Inpainting(tensor_size=img_size) # No mask for now
physics_generator = BernoulliSplittingMaskGenerator(tensor_size=img_size, split_ratio=0.5)

# Simulate a training loop for 2 epochs
out = []
for epoch in range(2):
    physics_generator.reset_rng()
    
    ys = []
    mask_prev = torch.tensor([0])
    for x in loader:
        params = physics_generator.step()
        y = physics(x, **params)
        ys += [y]
        # Test masks different every iteration
        assert not torch.equal(params["mask"], mask_prev)
        mask_prev = params["mask"]

    out += [torch.cat(ys, dim=-2)]

out = torch.cat(out, dim=-1)

# Test ys same every epoch
assert torch.equal(out[..., :8], out[..., 8:])

# Optional visualisation
plot(out)
```

minh-hai-nguyen-712 replied:
The issue #202 is solved now. To have the expected behavior, it's enough to call physiscs_generator.reset_rng() at the beginning of each epoch.

Andrewwango replied:
Ok, I added the appropriate warning to close #202

Andrewwango commented on deepinv/tests/test_physics.py:
```diff
@@ -580,13 +590,14 @@ def test_noise(device, noise_type):
     y1 = physics(
         x
-    )  # Note: this works but not physics.A(x) because only the noise is reset (A does not encapsulate noise)
+        # Note: this works but not physics.A(x) because only the noise is reset (A does not encapsulate noise)
```
And revert these stylistic changes caused by the formatter (and the lambda thing above)

Andrewwango commented on deepinv/physics/generator/inpainting.py:
```diff
@@ -301,14 +308,15 @@ def batch_step(self, input_mask: torch.Tensor = None) -> dict:
         for c in range(_C):
             for t in range(_T):
-                ind = self.rng.choice(
+                ind = random_choice(
```
Thank you so much for making these changes!!

Andrewwango commented on deepinv/physics/functional/random.py:
```diff
@@ -0,0 +1,70 @@
+import torch
+
+
+def random_choice(
```
I know it's a bit laborious, but please could you add a test for this function? Maybe just a hardcoded test to test the rng (i.e. `random_choice(...) == torch.Tensor([...])`, and something to test the replacement and p, like:
```python
a = torch.Tensor([1,2,3,4,5])
p = torch.Tensor([0,0,1,0,0])
random_choice(a, 2, replace=True, p=p) 
```
should equal to [3,3].

Andrewwango replied:
But thanks for providing this implementation! Is it taken from anywhere?

minh-hai-nguyen-712 replied:
I'll try for the tests. I implemented from scratch with the same logic and error as numpy, there is some examples provided in numpy docs for the case with equal probability. I'm not sure if there exists a better implementation out there.

Andrewwango commented:
@matthieutrs  please can you review the product blur convolution bits in this? I don't know anything about that!

matthieutrs commented on deepinv/physics/generator/mri.py:
```diff
@@ -19,22 +20,22 @@ class BaseMaskGenerator(PhysicsGenerator):
     :param int acceleration: acceleration factor, defaults to 4
     :param float center_fraction: fraction of lines to sample in low frequencies (center of k-space). If 0, there is no fixed low-freq sampling. Defaults to None.
-    :param np.random.Generator rng: numpy random generator, defaults to np.random.default_rng()
+    :param torch.Generator rng: torch random generator, defaults to None
```
```suggestion
    :param torch.Generator rng: torch random generator. Defaults to None.
```

matthieutrs commented on deepinv/tests/test_generators.py:
```diff
@@ -303,14 +323,778 @@ def test_generation(name, device):
+    elif name == "ProductConvolutionBlurGenerator":
+        w = params["filters"]
+        # atol = 1e-5
+        atol = 1e-2
```
Isn't this a bit too high?

minh-hai-nguyen-712 replied:
It passed on my 2 local machines with atol=1e-5 with newest torch version. But it doesn't pass on github.
I'm trying to figure out the problem. It would be great if you can test it on your local machine.

minh-hai-nguyen-712 replied:
It's weird for me. I tested it on multiple local machine and it passed every time. But it never passed on github.

mh-nguyen712 replied:
@matthieutrs Do you know on which hardware Github runs our tests? I ran the test on all of my local machine, I also tried on Google Colab and it passed every time.

mh-nguyen712 replied:
All test passed now. Some of PhysicsGenerator requires complex operation like torch.svd, which give different values depending on hardware, torch version, etc, so for the test, I only test on float64 with atol=1e-8.

matthieutrs replied:
Thanks!

matthieutrs commented on deepinv/tests/test_generators.py:
```diff
@@ -93,20 +112,21 @@ def test_generation_newparams(name, device):
 
-    params0 = generator.step(batch_size=batch_size)
-    params1 = generator.step(batch_size=batch_size)
+    params0 = generator.step(batch_size=batch_size, seed=0)
```
I did not find it either

minh-hai-nguyen-712 replied:
I'll add this test

matthieutrs reviewed (commented):
Thanks a lot for this amazing PR!

minh-hai-nguyen-712 commented on deepinv/physics/functional/interp.py:
```diff
@@ -179,15 +203,14 @@ def _radial_distance(self, X: torch.Tensor) -> torch.Tensor:
 
 
-def _ensure_2d(tensor: torch.Tensor) -> torch.Tensor:
+def _ensure_4d(tensor: torch.Tensor) -> torch.Tensor:
```
Yes your suggestion is better. I'll change it

minh-hai-nguyen-712 commented on deepinv/physics/generator/base.py:
```diff
@@ -26,41 +29,81 @@ class PhysicsGenerator(nn.Module):
+            if self.rng is not None:
+                self.rng = self.rng.manual_seed(seed)
+            else:
+                torch.manual_seed(seed)
```
Currently all randomness in our library come from torch. torch.manual_seed() will set the random seed for all devices. Your function is good for seed everything but It might cause some problems for other randomness? Also, in the __init__, if a rng is not provided, a new rng on the same device will be created under the hood so the call to torch.manual_seed() might disappear.

Andrewwango replied:
I agree with @mh-nguyen712 - I think our code should remain functional and not interfere with the global state - if the user really wants to set a universal seed maybe we can add the function that Matthieu suggested in `deepinv.utils` as a utility function?

minh-hai-nguyen-712 commented:
> Thanks a lot for this amazing PR!

Thank you very much for your detailed review!

Andrewwango commented:
Thanks for your work @mh-nguyen712 ! I have just added another issue #297  to also set rngs in physics which have randomness - if you are interested in combining with this PR!

mh-nguyen712 commented:
> Thanks for your work @mh-nguyen712 ! I have just added another issue #297 to also set rngs in physics which have randomness - if you are interested in combining with this PR!



> Thanks for your work @mh-nguyen712 ! I have just added another issue #297 to also set rngs in physics which have randomness - if you are interested in combining with this PR!

Of course! I can combine it to this PR!

mh-nguyen712 commented:
> Thanks for your work @mh-nguyen712 ! I have just added another issue #297 to also set rngs in physics which have randomness - if you are interested in combining with this PR!

I have added rng to all noise model, except for `GammaNoise`, which depends on torch.distributions and dont have support of `torch.Generator`. All tests passed. I also added some small code improvement (e.g., use new `random_choice` function). Could you review them?

Andrewwango commented:
I'm happy to approve - @matthieutrs ?

matthieutrs reviewed (commented):
Thanks for this great PR! Just reoppeing some minor comments which were closed but not taken into account it seems

matthieutrs commented on deepinv/physics/generator/base.py:
```diff
@@ -26,41 +29,81 @@ class PhysicsGenerator(nn.Module):
+        if rng is None:
+            self.rng = torch.Generator(device=device)
+        else:
+            # Make sure that the random generator is on the same device as the physic generator
```
@mh-nguyen712 I think this would actually be better with `physics`

mh-nguyen712 replied:
Sorry I only changed physic to physics in the warning message but forgot the comment.

mh-nguyen712 commented on deepinv/physics/noise.py:
```diff
@@ -177,11 +277,13 @@ class GammaNoise(torch.nn.Module):
     Distribution for modelling speckle noise (eg. SAR images),
     where :math:`\ell>0` controls the noise level (smaller values correspond to higher noise).
 
+    This noise model does not support the random number generator.
```
Added

mh-nguyen712 commented:
Sorry about that. I just resolved the open comments!
